### PR TITLE
Reading the directory from PATH and the symlink to this directory from PATH only once while create CommandCache

### LINKED ISSUE #3915: CommandsCache: reading the directory from PATH and the symlink to this directory from PATH (double work)
This is the second part of #3912

While I worked on #3911 I found that reading of the files in the PATH directories was done many times. 

I found that there is reading the directory from PATH and reading the symlink to this directory from PATH again. It makes xonsh start on Linux slower.

For example details by directories on Ubuntu:
```python
all=0
for p in $PATH:
    print(p)
    a=len([l for l in fp'{p}'.glob('*')])
    all += a
    print(a)
print(all)
```
```bash
/opt/miniconda/bin
353
/opt/miniconda/condabin
1
/home/pc/.local/bin
2
/usr/local/sbin
0
/usr/local/bin
0
/usr/sbin
420
/usr/bin
2298
/sbin
420
/bin
2298
/usr/games
0
/usr/local/games
0
/snap/bin
3

5795
```
And we can see that:
```bash
ls -la /bin
#/bin -> usr/bin

ls -la /sbin
#/sbin -> /usr/sbin
```
So xonsh scan the same directory twice.

As result instead of scan 5795-2298-420=3077 files once it scans 5795 files 3 times = 17,385 in sum.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### PR DESCRIPTION
**Pull request by anki-code**
Closes #3915

### CODE REVIEW

**Comment by scopatz** [comment]
I don't know know that this is a good idea. A lot of downstream environment management tools (like venv, conda, etc) assume that they can add and remove PATH elements. In general, it is assumed that PATH is pretty dumb. Thoughts @xonsh/xore?

**Comment by scopatz** [comment]
But thanks for the PR @anki-code!

**Comment by scopatz** [comment]
Maybe there is a way to make this optional and off by default?

**Comment by anki-code** [comment]
@scopatz please read the changes. I think you read only the title of the PR )

**Comment by scopatz** [comment]
Oh right, thanks.  Yeah this is on a different object than I thought

File: xonsh/commands_cache.py [hunk 1]
```diff
@@ -75,10 +75,20 @@ def remove_dups(p):
                 ret.append(e)
         return ret
 
+    @staticmethod
+    def remove_dups_path(p):
+        ret = list()
+        for e in p:
+            rpe = os.path.realpath(e)
+            e = e if e.lower() == rpe.lower() else rpe
```
Comment (scopatz) [inline xonsh/commands_cache.py:83]: I don't think lowering paths is supported cross platform

Comment (scopatz) [reply xonsh/commands_cache.py:83]: Maybe this should be a modification to `remove_dups()`. Is remove_dups used elsewhere?

Comment (anki-code) [reply xonsh/commands_cache.py:83]: It's only for checking and this line is needed exactly to cross platform - windows support.

Comment (scopatz) [reply xonsh/commands_cache.py:83]: The call site isn't guarded by an `if ON_WINDOWS` so this would break on Linux and other case-sensitive OSes

Comment (anki-code) [reply xonsh/commands_cache.py:83]: Hmm... but test fails...

Comment (anki-code) [reply xonsh/commands_cache.py]: OK, I avoided this.

**Comment by scopatz** [comment]
@anki-code - I think that this could be a one line change to the existing function. Here is how I would write it

```py
    @staticmethod
    def remove_dups(p):
        ret = list()
        for e in map(os.path.realpath, p):
            if e not in ret:
                ret.append(e)
        return ret
```

**Comment by scopatz** [comment]
This looks good to me now! Let me know when you think it is ready!

**Comment by anki-code** [comment]
I think it's ready )

**Pull request merged**